docs/adr/reconstruction.md

```diff
@@ -0,0 +1,286 @@
+1. High frequency of duplicate requests leading to network inefficiency
+2. Suboptimal bandwidth utilization
+3. Limited scalability with increasing block sizes and node count
+
+Key improvements include:
+- Structured bitmap sharing
+- Optimized proof packaging
+- Efficient state management
+- Robust failure handling
+
+This proposal aims to implement an efficient reconstruction protocol that:
+- Minimizes duplicate data requests
+- Optimizes bandwidth usage through smart data packaging
+- Supports network scaling across multiple dimensions
+- Maintains stability under varying network conditions
+
+Engineering time.
+
+Initial draft  aims to be optimised in terms of engineering efforts required for first iteration of implementation. Optimizations marked as optional can be implemented in subsequent updates based on network performance metrics.
+
+## Specification
+
+### Performance Requirements
+
+1. Base Scenario Support:
+    - 32MB block size
+    - Minimum required light nodes for 32MB blocks
+    - Network of 50+ full nodes
+
+2. Performance Metrics (in order of priority):
+    - System stability
+    - Reconstruction throughput (blocks/second)
+    - Per-block reconstruction time
+
+3. Scalability Dimensions:
+    - Block size scaling
+    - Node count scaling
+
+
+
+### Protocol Flow
+
+```
+1. Connection Establishment
+   Client                              Server
+      |---- Handshake (node type) ----->|
+      |<---- Handshake (node type) -----|
+
+2. Bitmap Subscription
+   Client                                   Server
+      |---- Subscribe to bitmap ------------->|
+      |<---- Initial bitmap ------------------|
+      |<---- Updates  ------------------------|
+      |<---- End updates(full eds/max samples-|
+
+3. Data Request
+   Client                              Server
+      |---- Request(bitmap) ----------->|
+      |<---- [Samples + Proof] parts ---|
+```
+
+### Bitmap Implementation
+
+The protocol utilizes Roaring Bitmaps for efficient bitmap operations and storage. Roaring Bitmaps provide several advantages for the reconstruction protocol:
+
+1. Efficient Operations
+    - Fast logical operations (AND, OR, XOR)
+    - Optimized for sparse and dense data
+    - Memory-efficient storage
+
+2. Implementation Benefits
+    - Native support for common bitmap operations
+    - Optimized serialization
+    - Efficient iteration over set bits
+    - Support for rank/select operations
+
+3. Performance Characteristics
+    - O(1) for most common operations
+    - Compressed storage format
+    - Efficient memory utilization
+    - Fast bitmap comparisons
+
+
+### Core Protocol Components
+
+#### 1. Connection Management
+
+Handshakes, Node Type Identification
+- Node type must be declared during connection handshake by each of connection nodes
+- Active connection state maintenance on full nodes. List should be maintained by node type
+- Connection state should allow subscription for updated
```

Comment:
Whats the rationale behind sharing your node type? Is it possible to avoid sharing this information with peers?

There is a privacy concern we discussed a while ago with @musalbas and @adlerjohn. FNs ideally should be indistinguishable from LNs on p2p network, so they validators don't detect them.

If we don't think that's important anymore or sharing type is unavoidable, I also propose to cut the handshake roundtrip and rely on libp2p toolset for that.

Reply:
That is indeed a goal, but the problem here is that it's implicit in this protocol that only full nodes request bitmaps from others, regardless of how you identify in the handshake.

FNs pretending to be LNs (to reconstruct) would only work if the FNs spam sample requests to the block producer, which isn't something implemented yet afaik. (See https://forum.celestia.org/t/security-levels-for-data-availability-for-light-nodes/919#level-3-data-availability-sampling-without-an-honest-minority-of-light-nodes-6)

Reply:
> if the FNs spam sample requests to the block producer, which isn't something implemented yet afaik. 

Do you think this should be in scope of current iteration of reconstruction protocol?

Reply:
I think not in this iteration, because it seems we're doing pretty good on increasing number of light nodes, so we could just focus directly on having a robust construction protocol and achieving Level 4 light nodes. It might be nice to have later as an option, though, but not high priority right now.

Reply:
Good point, handshake is not necessary for initial basic implementation. It will be needed in peer scoring optimisation. When FN would need to know the type of peer for prioritisation of requests. I'll mark handshake protocol as future optimisation.